support_opening_files_through_url_handling_in_vscode/README.md

```diff
@@ -0,0 +1,66 @@
+# Support opening files through URL handling in Vscode
```

Comment:
Cool you found places in vscode to alter.

Any particular reason you want to modify vscode to support the protocol directly at this stage ? 

One of the ideas/values behind starfish is that each IDE (at least for basic operations) should *not* need to support the protocol directly. Thoughts ?

Reply:
The reason for this was to understand the functionality of vscode.

I agree that we dont need to modify the code directly for vscode.There can be better solutions.
One of them could be through Extensions using "Extension API" of vscode.

And if we've got any better choice than an extension I can start working on that as well.

Reply:
Additionally I saw the command line tool of vscode :[code.visualstudio.com](https://code.visualstudio.com/docs/editor/command-line).

With this we can easily execute vscode commands ,as stated on [code.visualstudio.com](https://code.visualstudio.com/docs/editor/command-line) "Visual Studio Code has a powerful command line interface built-in that lets you control how you launch the editor. You can open files, install extensions, change the display language, and output diagnostics through command-line options (switches)"

Reply:
Also the mapping of URL to Application (i.e Starfish) can be done as done here: [https://github.com/shengyou/vscode-handler](https://github.com/shengyou/vscode-handler)

For mapping, we have to execute `.reg` Files(_on Windows_) or execute `bash install` on _Linux_ 
We can achieve this easily by automating our app to execute intended codes from _command line_

Reply:
Yes, vscode has url extensions too and a nice cli - the downside being it only works for vscode.

That’s why starfish would be its own independent script/executable which we register in the different OS/browsers url handler logic and then can make it work for basically any IDE.

Reply:
So is my reasearch ample enough to start the project or shall i reasearch more before starting the implementation

Reply:
if you want to do it as part of GSOC you'll make a proposal (https://google.github.io/gsocguides/student/writing-a-proposal) - for now we submit those at https://docs.google.com/forms/d/1ePMj3FRBQ_BNYo9pKSaYI9HE0uv_gTLksJWqSOqYdIw/prefill

Reply:
Thanks, Max !!  I'll submit a proposal soon.

Meanwhile, I've updated the URL Handlers for Linux

Reply:
Update : I've started to work with [navigator.registerProtocolHandler(protocol, url, title);](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler) for protocol handling from Browser that would be routed to xdg-open....

Reply:
cool, but just keep it at "ide://"

Reply:
Sure I'll keep it at `ide://`

Meanwhile, I was trying to figure out the development of the **Starfish App**.
Should we use **Electron JS** to build the application and keep it, **user interactive**?

Reply:
Hey I've managed to get a mockup of Open in Starfish:

Visit here: https://fahad-israr.github.io/starfish-demo/

# starfish-demo
Demonstrates open in IDE Option on Github


_As a mockup we open `apt` instead of starfish(as it is not developed yet)_

- on site load it prompts to add url handler extension:site wants to *`open web+apt links`* with two options *`block`* and *`allow`*
- click allow to register apt
- now click on open in starfish to launch apt
- it will prompt to `open xdg-open`
- click on `open-xdg-open` 
- note that we are trigerring apt with the url for mockup 
- ofcourse it is an invalid package so t will give error

Reply:
site doesn't load for me.

but cool :)

Reply:
Thats because you are using windows or Mac OS (Probably)....The app  I am launching is apt that comes with Linux.....I've done it for testing purpose....Could you please try opening on Linux....And please elaborate the issue you faced... I'll try resolving it!!

Reply:
Once we install our starfish url handler on all machines this would start working probably...coz then we would launch Starfish Registered on all three platforms..

Reply:
And please help me on this  as well https://github.com/maxandersen/starfish/pull/1#discussion_r390707239

Reply:
seemed to just been a temporary network issue. now it loads and just tells me I don't have any apps that react to the starfish protocol - which is true

Reply:
> Sure I'll keep it at `ide://`
> 
> Meanwhile, I was trying to figure out the development of the **Starfish App**.
> Should we use **Electron JS** to build the application and keep it, **user interactive**?

I don't really (yet) see a usecase for having it much interactive.  What do you see need for interaction to begin with ? 

its supposed to just do the right thing and open up the users IDE...

Reply:
> seemed to just been a temporary network issue. now it loads and just tells me I don't have any apps that react to the starfish protocol - which is true

Thats awesome!!!

Reply:
> > Sure I'll keep it at `ide://`
> > Meanwhile, I was trying to figure out the development of the **Starfish App**.
> > Should we use **Electron JS** to build the application and keep it, **user interactive**?
> 
> I don't really (yet) see a usecase for having it much interactive. What do you see need for interaction to begin with ?
> 
> its supposed to just do the right thing and open up the users IDE...

- User can have multiple IDE...How will he configure which one to open....?

- Also, as we are going to develop extensions for IDEs it will be tedious for users to go to each IDE and choose and Install extension to support Starfish. Instead we will display  all installed IDEs and show option to install our extension.

- Incase user encounter issues he can troubleshoot using help guide we provide with the starfish  app or report to us easily incase unresolved.

- So i think a user friendly GUI can be quite helpful

Reply:
> * User can have multiple IDE...How will he configure which one to open....?

to begin with start with a configuration file.

> * Also, as we are going to develop extensions for IDEs it will be tedious for users to go to each IDE

what usecases do you need/want the extensions for ? if we had some these could literally just be on the home page of starfish (at least to begin with - keep it simple to start)

> * Incase user encounter issues he can troubleshoot using help guide we provide with the starfish  app or report to us easily incase unresolved.

logs  or a basic webui could suffice for that.

> So i think a user friendly GUI can be quite helpful

I follow what you are saying but here is my thinking: 
the more layers of ui we show the user the worse he is off - the absolute best experience would be to just open up the IDE and not see any UI - if we start with the outset of adding UI we'll not solve the actual main issue. I suggest we focus on getting the basics in place and add layers on it.

i..e configuration UI does not need a desktop app - it could literally just be a embedded webapp.

also the smaller we can make starfish the better - electron apps are massive; even the small ones.

Reply:
Thanks for clarification Max. I'll adhere to it and work accordingly.

Reply:
since I work on quarkus and would love to see a native binary for this or at least a runtime that is easily bundled I would prefer its written in java using jbang or quarkus command-mode ....the latter at least would allow us to easily expose a webui (which could be written in whaterver - i'm mainly for keeping it super simple for the user).

anything node or python bound would require additional setup. java or native is something that can be shipped fairly lightweight.

thats my thinking - whats yours ?

Reply:
It's cool...we'll build on [Quarkus](https://quarkus.io/) then!!
But I don't have any experience working with [Quarkus](https://quarkus.io/). As a tech enthusiast, I love to learn new technologies. I am sure I will be able to work with it soon with your guidance.
Currently, I am following:"[QUARKUS - BUILDING A NATIVE EXECUTABLE](https://quarkus.io/guides/building-native-image).
Please do refer to me the resources you may think might be helpful for improved learning.

Reply:
> if you want to do it as part of GSOC you'll make a proposal (https://google.github.io/gsocguides/student/writing-a-proposal) - for now we submit those at https://docs.google.com/forms/d/1ePMj3FRBQ_BNYo9pKSaYI9HE0uv_gTLksJWqSOqYdIw/prefill

The docs need access.I've requested access. Please check it.

Reply:
granted. din't realize these were publically limited. but lets use it for now.

Reply:
> Please do refer to me the resources you may think might be helpful for improved learning.

quarkus.io/guides is a good start.

mind you that most material in quarkus is about using it for services (i.e. microservices) the notion of command mode for use in cli's and batch jobs is coming in 1.4 (a month from now) so to start just doing basic java should be fine.

but don't let that stop you from exploring Quarkus today! :)

Reply:
> granted. din't realize these were publically limited. but lets use it for now.

Thanks!!

Reply:
> > Please do refer to me the resources you may think might be helpful for improved learning.
> 
> quarkus.io/guides is a good start.
> 
> mind you that most material in quarkus is about using it for services (i.e. microservices) the notion of command mode for use in cli's and batch jobs is coming in 1.4 (a month from now) so to start just doing basic java should be fine.
> 
> but don't let that stop you from exploring Quarkus today! :)

That's awesome!! I'd keep exploring [Quarkus](https://quarkus.io/) and start developing with it as soon as possible!!

Reply:
> > if you want to do it as part of GSOC you'll make a proposal (https://google.github.io/gsocguides/student/writing-a-proposal) - for now we submit those at https://docs.google.com/forms/d/1ePMj3FRBQ_BNYo9pKSaYI9HE0uv_gTLksJWqSOqYdIw/prefill
> 
> The docs need access.I've requested access. Please check it.

Hi max,I've successfully submitted the Google form.
Please review it and give me suggestions to improve it.

Reply:
hmm - I haven't received any notfication and nothing listed in the results tab - can you double check you pressed submit ? and did you send the pdf to google's gsoc system ?

Reply:
I am sorry...i forgot to press submit button after prefilling.
You may review my draft now.

Reply:
Hi max,
did you go through my submitted draft?